plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,157 @@
+   belons to is running out of space, nodeos will crash with SIGBUS as
+   the state file is unable to acquire new pages.
+
+   The solution is to have a dedicated plugin to monitor resource 
+   usages (file system space now, CPU, memory, and networking
+   bandwidth in the future).
+   The plugin uses a thread to periodically check space usage of file
+   systems of directories being monitored. If space used
+   is over a predefined threshold, a graceful shutdown is initiated.
+**/
+
+#include <eosio/resource_monitor_plugin/resource_monitor_plugin.hpp>
+#include <eosio/resource_monitor_plugin/file_space_handler.hpp>
+#include <eosio/resource_monitor_plugin/system_file_space_provider.hpp>
+
+#include <eosio/chain/exceptions.hpp>
+
+#include <fc/exception/exception.hpp>
+#include <fc/log/logger_config.hpp> // set_os_thread_name
+
+#include <boost/filesystem.hpp>
+
+#include <thread>
+
+#include <sys/stat.h>
+
+using namespace eosio::resource_monitor;
+
+namespace bfs = boost::filesystem;
+
+namespace eosio {
+   static appbase::abstract_plugin& _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
+
+   static vector<bfs::path> directories_monitored;
+
+class resource_monitor_plugin_impl {
+public:
+   resource_monitor_plugin_impl()
+      :space_handler(system_file_space_provider(), ctx)
+      {
+      }
+
+   void set_program_options(options_description&, options_description& cfg) {
+      cfg.add_options()
+         ( "resource-monitor-interval-seconds", bpo::value<uint32_t>()->default_value(def_sleep_time_in_secs),
+           "Time in seconds between two consecutive checks of resource usage. Should be greater than 1 and less than 60 (one minute)" )
+         ( "resource-monitor-space-threshold", bpo::value<uint32_t>()->default_value(def_used_space_threshold_in_percentage),
+             "Threshold in terms of percentage of used space vs total space. If used space is over the threshold, a graceful shutdown is initiated. The value should be greater than 1 to less than 99" )
+         ;
+   }
+
+   void plugin_initialize(const appbase::variables_map& options) {
+      ilog("plugin_initialize");
+   
+      auto sleep_time= options.at("resource-monitor-interval-seconds").as<uint32_t>();
+      EOS_ASSERT(sleep_time> 0 && sleep_time < 60, chain::plugin_config_exception,
+         "\"resource-monitor-interval-seconds\" must be greater than 0 and less than 60 (1 minute)");
+
+      space_handler.set_sleep_time(sleep_time);
+      ilog("sleep_time = ${sleep_time}", ("sleep_time", sleep_time));
+
+      auto threshold = options.at("resource-monitor-space-threshold").as<uint32_t>();
+      EOS_ASSERT(threshold > 0 && threshold < 100, chain::plugin_config_exception,
+         "\"resource-monitor-space-threshold\" must be greater than 0 and less than 100");
+
+      space_handler.set_threshold(threshold);
+      ilog("threshold = ${threshold}", ("threshold", threshold));
```

Comment:
This should be more descriptive if an info level log or made debug level.

Reply:
Will make it more descriptive.